### ISSUE TITLE
endpoints can get into not-ready state upon restoring when regenerations are triggered after endpoints are inserted into endpointmanager, but before restoration has occurred

### ISSUE THREAD

**Issue by aanm** [cilium/cilium#8234]
```
Failed while regenerating endpoint
```
[0aaea1be_K8sFQDNTest_Restart_Cilium_validate_that_FQDN_is_still_working.zip](https://github.com/cilium/cilium/files/3261708/0aaea1be_K8sFQDNTest_Restart_Cilium_validate_that_FQDN_is_still_working.zip)

**Comment by ianvernon** [comment]
@aanm do you have a build link?

**Comment by aanm** [comment]
@ianvernon https://jenkins.cilium.io/job/Cilium-PR-Ginkgo-Tests-K8s/2182/testReport/junit/Suite-k8s-1/12/K8sFQDNTest_Restart_Cilium_validate_that_FQDN_is_still_working/

**Comment by ianvernon** [comment]
Thanks !

**Comment by ianvernon** [comment]
The reason behind the failure is the following:

```
2019-06-06T10:34:24.18864636Z level=debug msg="Skipping build due to invalid state" containerID=0451a66fd1 datapathPolicyRevision=4 desiredPolicyRevision=4 endpointID=2859 endpointState=ready identity=51099 ipv4=10.10.0.182 ipv6="f00d::a0a:0:0:2446" k8sPodName=default/app2-544fbfcbbb-d6mqk subsys=endpoint
2019-06-06T10:34:24.18866236Z level=warning msg="Regeneration of endpoint failed" bpfCompilation=0s bpfLoadProg=0s bpfWaitForELF=0s bpfWriteELF=0s buildDuration="92.005µs" containerID=0451a66fd1 datapathPolicyRevision=4 desiredPolicyRevision=4 endpointID=2859 error="Skipping build due to invalid state: ready" identity=51099 ipv4=10.10.0.182 ipv6="f00d::a0a:0:0:2446" k8sPodName=default/app2-544fbfcbbb-d6mqk mapSync=0s policyCalculation=0s prepareBuild=0s proxyConfiguration=0s proxyPolicyCalculation=0s proxyWaitForAck=0s reason="syncing state to host" subsys=endpoint waitingForCTClean=0s waitingForLock="4.32µs"
2019-06-06T10:34:24.188669333Z level=debug msg="Error regenerating endpoint: Skipping build due to invalid state: ready" code=Failure containerID=0451a66fd1 datapathPolicyRevision=4 desiredPolicyRevision=4 endpointID=2859 endpointState=ready identity=51099 ipv4=10.10.0.182 ipv6="f00d::a0a:0:0:2446" k8sPodName=default/app2-544fbfcbbb-d6mqk policyRevision=4 subsys=endpoint type=200
2019-06-06T10:34:24.188674364Z level=error msg="endpoint regeneration failed" containerID=0451a66fd1 datapathPolicyRevision=4 desiredPolicyRevision=4 endpointID=2859 error="Skipping build due to invalid state: ready" identity=51099 ipv4=10.10.0.182 ipv6="f00d::a0a:0:0:2446" k8sPodName=default/app2-544fbfcbbb-d6mqk subsys=endpoint
```

So we are trying to regenerate an endpoint which is already in `ready` state. But, there doesn't appear to be anything wrong with the endpoint per-se at a first glance.

**Comment by ianvernon** [comment]
The failed regeneration which was triggered was due to `syncing state to host`. The only time such a reason appears in the code is in `daemon/state.go:regenerateRestoredEndpoints`:

https://github.com/cilium/cilium/blob/ec9019ac1961f865bc0e280cd1b36bce0bcb50df/daemon/state.go#L351-L366

Given that we unlock the endpoint after this state change, it's entirely possible that we change the state elsewhere for the same endpoint, as the endpoint is exposed to the endpoint manager right above:

https://github.com/cilium/cilium/blob/ec9019ac1961f865bc0e280cd1b36bce0bcb50df/daemon/state.go#L230-L256

**Comment by ianvernon** [comment]
We insert the endpoint into the endpoint manager at `10:34:16`:

```
2019-06-06T10:34:16.120545166Z level=info msg="New endpoint" containerID=0451a66fd1 datapathPolicyRevision=0 desiredPolicyRevision=0 endpointID=2859 identity=51099 ipv4=10.10.0.182 ipv6="f00d::a0a:0:0:2446" k8sPodName=default/app2-544fbfcbbb-d6mqk subsys=endpoint
```

Right after this happens, we trigger endpoint regeneration due to policy rules being added. This can happen, as the endpoint is exposed to the rest of the agent via the endpointmanager. 

```
2019-06-06T10:34:16.531113493Z level=debug msg="Triggering endpoint regeneration due to policy rules added" code=OK containerID=0451a66fd1 datapathPolicyRevision=0 desiredPolicyRevision=0 endpointID=2859 endpointState=waiting-to-regenerate identity=51099 ipv4=10.10.0.182 ipv6="f00d::a0a:0:0:2446" k8sPodName=default/app2-544fbfcbbb-d6mqk policyRevision=0 subsys=endpoint type=0
```

Then, as part of the restore logic, we then set the state back to `restoring` with the "Synchronizing endpoint labels with KVStore" reason:

```
2019-06-06T10:34:16.559918549Z level=debug msg="Synchronizing endpoint labels with KVStore" code=OK containerID=0451a66fd1 datapathPolicyRevision=0 desiredPolicyRevision=0 endpointID=2859 endpointState=restoring identity=51099 ipv4=10.10.0.182 ipv6="f00d::a0a:0:0:2446" k8sPodName=default/app2-544fbfcbbb-d6mqk policyRevision=0 subsys=endpoint type=0
```

Given that the endpoint is exposed at this point via endpointmanager, we have no guarantees about endpoint state in the restoring process. This means that we can try to regenerate an endpoint while it is currently regenerating, and sometimes this means that we can try to transition into an invalid state while regenerating in the goroutine managing the restoring of the endpoint. 

I suspect that the reason this manifests during the FQDN restart test is that the FQDN subsystem adds policy rules into the repository, and it may trigger regeneration of endpoints before or after they are inserted into the endpointmanager, explaining the intermittence of this failure.

**Comment by ianvernon** [comment]
I'm unsure about how to fix this as of now, as our bootstrapping for endpoints is quite complicated. I feel like it would be best for us to not regenerate endpoints until the FQDN policies that apply to them are actually resolved via restoration of their DNS Caches. 

cc: @raybejjani . What do you think?

**Comment by ianvernon** [comment]
I think that https://github.com/cilium/cilium/pull/8117 will actually help with this issue in terms of how the CI is affected, but it won't help with underlying problem; we are exposing endpoints via endpointmanager too early.

**Comment by ianvernon** [comment]
This was fixed by #8551
